bindings/imap-codec-python/src/lib.rs

```diff
@@ -1,15 +1,244 @@
+        Ok(PyEncoded(Some(encoded)))
+    }
+}
+
+/// Wrapper for `ResponseCodec`
+#[derive(Debug, Clone, PartialEq)]
+#[pyclass(name = "ResponseCodec")]
+struct PyResponseCodec(ResponseCodec);
+
+#[pymethods]
+impl PyResponseCodec {
+    /// Decode response from given bytes
+    #[staticmethod]
+    fn decode(bytes: Bound<PyBytes>) -> PyResult<(Bound<PyBytes>, Bound<PyAny>)> {
+        let py = bytes.py();
+        match ResponseCodec::default().decode(bytes.as_bytes()) {
+            Ok((remaining, response)) => Ok((
+                PyBytes::new_bound(py, remaining),
+                serde_pyobject::to_pyobject(py, &response)?,
+            )),
+            Err(err) => Err(match err {
+                decode::ResponseDecodeError::Incomplete => DecodeIncomplete::new_err(()),
+                decode::ResponseDecodeError::LiteralFound { length } => {
+                    let dict = pyo3::types::PyDict::new_bound(py);
+                    dict.set_item("length", length)?;
+                    DecodeLiteralFound::new_err(dict.unbind())
+                }
+                decode::ResponseDecodeError::Failed => DecodeFailed::new_err(()),
+            }),
+        }
+    }
+
+    /// Encode response into fragments
+    #[staticmethod]
+    fn encode(response: Bound<PyAny>) -> PyResult<PyEncoded> {
+        let response = serde_pyobject::from_pyobject(response)?;
+        let encoded = ResponseCodec::default().encode(&response);
+        Ok(PyEncoded(Some(encoded)))
+    }
+}
+
+/// Wrapper for `IdleDoneCodec`
+#[derive(Debug, Clone, PartialEq)]
+#[pyclass(name = "IdleDoneCodec")]
+struct PyIdleDoneCodec(IdleDoneCodec);
+
+#[pymethods]
+impl PyIdleDoneCodec {
+    /// Decode idle done from given bytes
+    #[staticmethod]
+    fn decode(bytes: Bound<PyBytes>) -> PyResult<(Bound<PyBytes>, Bound<PyAny>)> {
+        let py = bytes.py();
+        match IdleDoneCodec::default().decode(bytes.as_bytes()) {
+            Ok((remaining, idle_done)) => Ok((
+                PyBytes::new_bound(py, remaining),
+                serde_pyobject::to_pyobject(py, &idle_done)?,
+            )),
+            Err(err) => Err(match err {
+                decode::IdleDoneDecodeError::Incomplete => DecodeIncomplete::new_err(()),
+                decode::IdleDoneDecodeError::Failed => DecodeFailed::new_err(()),
+            }),
+        }
+    }
+
+    /// Encode idle done into fragments
+    #[staticmethod]
+    fn encode(idle_done: Bound<PyAny>) -> PyResult<PyEncoded> {
+        let idle_done = serde_pyobject::from_pyobject(idle_done)?;
+        let encoded = IdleDoneCodec::default().encode(&idle_done);
+        Ok(PyEncoded(Some(encoded)))
+    }
 }
 
-/// A Python module implemented in Rust.
 #[pymodule]
 #[pyo3(name = "imap_codec")]
 fn imap_codec_python(m: &Bound<'_, PyModule>) -> PyResult<()> {
-    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;
+    m.add_class::<PyEncoded>()?;
+    m.add_class::<PyGreetingCodec>()?;
```

Comment:
You're welcome. When I know somethings has a deliberate order, I'm happy to keep that. Otherwise, I default to alphabetically to reduce thinking.
I now see, that I'm a bit inconsistent about where `Encoded` and the error types are sorted, but well ...